File: 2140-solving-questions-with-brainpower/2140-solving-questions-with-brainpower.py
class Solution(object):
    def mostPoints(self, questions):
        """
        :type questions: List[List[int]]
        :rtype: int
        """
        n = len(questions)
        v = [0] * n

        def solve(i):
            if i >= n: return 0
                    
            if v[i]: return v[i]
            
            count = questions[i][0] + solve(i+1 + questions[i][1])
            
            skip = solve(i+1)
            
            v[i] = max(count, skip)
            return v[i]
        
        return solve(0)

File: 2140-solving-questions-with-brainpower/2140-solving-questions-with-brainpower.java
class Solution {
    int n;
    int[][] question;
    long[] v; 
    long solve(int i){
        if(i >= n) return 0;
        if(v[i]!=0)
            return v[i];
        //count
        long count = (question[i][0] + solve(i + question[i][1] + 1));
        
        //skip
        long skip = solve(i+1);
        
        return v[i] = Math.max(count, skip);
    }
    
    public long mostPoints(int[][] questions) {
        n = questions.length;
        question = questions;
        v = new long[n]; 
        return solve(0);
    }
}